dashboard/modules/serve/serve_agent.py

```diff
@@ -74,6 +74,36 @@ async def get_all_deployments(self, req: Request) -> Response:
             content_type="application/json",
         )
 
+    @routes.get("/api/serve/applications/")
+    @optional_utils.init_ray_and_catch_exceptions()
+    async def get_serve_instance_details(self, req: Request) -> Response:
```

Comment:
Does this have to be in the serve_agent.py? Can this be in a `serve_head.py` file instead?

I believe the main difference is there is one agent process per node in the cluster, but for head, there is only one head process and its on the head node.

If this is primarily used for reading data from the serve actor, does every agent need to serve this data or can it just be served from the head node?

This would be more aligned with how the other dashboard apis are implemented. (state-api, jobs apis, etc)

Reply:
@shrekris-anyscale can help provide insight here?

Reply:
Let's keep this in the agent for now. We should raise this with others on the Serve team, and potentially move the Serve REST API as a whole in a separate PR.

For context, the Serve REST API used to be on the head node under a `serve_head.py` file. However, it was moved to the dashboard agent to provide better fault tolerance in case the head node (and the dashboard) crashes. Any higher-level service executor (e.g. KubeRay) needs access to the REST API to check Serve's health. If the API is in the head node, then anytime the head node goes down, Serve's health check would fail, and the executor would restart Serve instead of waiting for it to recover. By putting the REST API in the dashboard agent, the executor can check Serve's health at other cluster nodes while the head node is recovering.